x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherService.java

```diff
@@ -144,24 +146,31 @@ public boolean validate(ClusterState state) {
     }
 
     /**
-     * Stops the watcher service and marks its services as paused
+     * Stops the watcher service and marks its services as paused. Callers should set the Watcher state to {@link WatcherState#STOPPING}
+     * prior to calling this method.
+     *
+     * @param stoppedListener The listener that will set Watcher state to: {@link WatcherState#STOPPED}, may not be {@code null}
      */
-    public void stop(String reason) {
+    public void stop(String reason, Consumer<Void> stoppedListener) {
+        assert stoppedListener != null;
         logger.info("stopping watch service, reason [{}]", reason);
-        executionService.pause();
+        executionService.pause(stoppedListener);
         triggerService.pauseExecution();
     }
 
     /**
      * shuts down the trigger service as well to make sure there are no lingering threads
      * also no need to check anything, as this is final, we just can go to status STOPPED
+     *
+     * @param stopListener The listener that will set Watcher state to: {@link WatcherState#STOPPED}, may be {@code null} assuming the
+     *                     {@link WatcherState#STOPPED} is set elsewhere or not needed to be set.
      */
-    void shutDown() {
+    void shutDown(@Nullable Consumer<Void> stopListener) {
```

Comment:
consistent naming: stopListener vs. stoppedListener

Reply:
done in https://github.com/elastic/elasticsearch/pull/43939/commits/39b0336b38555d3a2590c063068cd595ee327210